### PR TITLE
fix: stop phpunit premature exits

### PR DESCRIPTION
**Pull request by superdav42**
## Summary

- Intercept AJAX/JSON `wp_die()` paths in affected PHPUnit tests so `wp_send_json_*()` no longer exits the runner early.
- Replace non-interceptable raw gateway exits/status handling with WordPress JSON/die responses where needed.
- Harden test fixtures around PayPal OAuth namespace mocks, malformed multisite globals, and `WP_Admin_Bar` mock loading so the full suite reaches PHPUnit's normal summary.

## Verification

- `vendor/bin/phpcs inc/functions/fs.php inc/functions/gateway.php inc/gateways/class-paypal-oauth-handler.php tests/WP_Ultimo/Gateways/PayPal_OAuth_Handler_Test.php tests/WP_Ultimo/Gateways/PayPal_OAuth_Handler_Standalone_Test.php`
- `vendor/bin/phpcs inc/managers/class-gateway-manager.php tests/WP_Ultimo/Managers/Gateway_Manager_Test.php`
- `vendor/bin/phpcs tests/WP_Ultimo/Admin_Pages/Top_Admin_Nav_Menu_Test.php`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter PayPal_OAuth_Handler_Standalone_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter PayPal_OAuth_Handler_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter Cloudflare_Host_Provider_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter Gateway_Manager_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter Top_Admin_Nav_Menu_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter Whitelabel_Test`
- `set -o pipefail; XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --debug > /tmp/opencode/ultimate-multisite-phpunit-debug.log 2>&1; status=$?; printf 'PHPUNIT_EXIT_STATUS=%s\nLOG=/tmp/opencode/ultimate-multisite-phpunit-debug.log\n' "$status"; exit "$status"`

## Notes

- The full suite now completes to a normal PHPUnit summary instead of fataling/exiting prematurely.
- Current full-suite summary is still red and needs follow-up: `Tests: 10001, Assertions: 25066, Errors: 545, Failures: 120, Skipped: 57, Risky: 2.`
- Pre-commit hooks passed for the committed production PHP files: PHPCS and PHPStan.

---
[aidevops.sh](https://aidevops.sh) v3.21.11 plugin for [OpenCode](https://opencode.ai) v1.17.9 with gpt-5.5

## Summary by CodeRabbit

# Release Notes

* **Bug Fixes**
  * Improved PayPal webhook failure/pending handling with correct termination and HTTP status behavior.
  * Applied the gateway filter consistently when a gateway is missing or available.
  * Enhanced multisite main-site upload directory resolution and safer state handling.
  * Added validation to prevent editing an item when its payment record is missing.
  * Made template “more info” rendering stop early when no matching template is found.

* **Tests**
  * Reworked AJAX/JSON response assertions to capture `wp_send_json_*` payloads and redirect targets more reliably.
  * Expanded and adjusted SSO coverage around redirect and edge-case handling.

### CODE REVIEW

**Comment by superdav42** [comment]
## Summary
- Intercept AJAX/JSON `wp_die()` paths in affected PHPUnit tests so `wp_send_json_*()` no longer exits the runner early.
- Replace non-interceptable raw gateway exits/status handling with WordPress JSON/die responses where needed.
- Harden test fixtures around PayPal OAuth namespace mocks, malformed multisite globals, and `WP_Admin_Bar` mock loading so the full suite reaches PHPUnit's normal summary.
## Verification
- `vendor/bin/phpcs inc/functions/fs.php inc/functions/gateway.php inc/gateways/class-paypal-oauth-handler.php tests/WP_Ultimo/Gateways/PayPal_OAuth_Handler_Test.php tests/WP_Ultimo/Gateways/PayPal_OAuth_Handler_Standalone_Test.php`
- `vendor/bin/phpcs inc/managers/class-gateway-manager.php tests/WP_Ultimo/Managers/Gateway_Manager_Test.php`
- `vendor/bin/phpcs tests/WP_Ultimo/Admin_Pages/Top_Admin_Nav_Menu_Test.php`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter PayPal_OAuth_Handler_Standalone_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter PayPal_OAuth_Handler_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter Cloudflare_Host_Provider_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter Gateway_Manager_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter Top_Admin_Nav_Menu_Test`
- `XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --filter Whitelabel_Test`
- `set -o pipefail; XDEBUG_MODE=off vendor/bin/phpunit --no-coverage --debug > /tmp/opencode/ultimate-multisite-phpunit-debug.log 2>&1; status=$?; printf 'PHPUNIT_EXIT_STATUS=%s\nLOG=/tmp/opencode/ultimate-multisite-phpunit-debug.log\n' "$status"; exit "$status"`
## Notes
- The full suite now completes to a normal PHPUnit summary instead of fataling/exiting prematurely.
- Current full-suite summary is still red and needs follow-up: `Tests: 10001, Assertions: 25066, Errors: 545, Failures: 120, Skipped: 57, Risky: 2.`
- Pre-commit hooks passed for the committed production PHP files: PHPCS and PHPStan.

---
[aidevops.sh](https://aidevops.sh) v3.21.11 plugin for [OpenCode](https://opencode.ai) v1.17.9 with gpt-5.5

---
Merged via PR #1471 to main.
_Merged by deterministic merge pass (pulse-wrapper.sh)._

File: tests/WP_Ultimo/Admin_Pages/Customer_List_Admin_Page_Test.php [hunk 1]
```diff
@@ -418,10 +486,15 @@ public function test_handle_add_new_customer_modal_new_type_with_valid_email():
...
+
+		$this->assertArrayHasKey('success', $response);
 
-		$this->page->handle_add_new_customer_modal();
+		if ($response['success']) {
+			$this->assertArrayHasKey('redirect_url', $response['data']);
+			$this->assertStringContainsString('wp-ultimo-edit-customer', $response['data']['redirect_url']);
+		} else {
+			$this->assertIsArray($response['data']);
+		}
```
[the remark this answers is not in the document]
Comment (superdav42) [reply tests/WP_Ultimo/Admin_Pages/Customer_List_Admin_Page_Test.php:498]: @coderabbitai Addressed in dfc8a34e: the fallback path now asserts `success` is false instead of only checking `data` shape (`tests/WP_Ultimo/Admin_Pages/Customer_List_Admin_Page_Test.php`). Verified with focused PHPUnit: 303 tests, 564 assertions, 4 skipped.

File: tests/WP_Ultimo/Admin_Pages/Payment_Edit_Admin_Page_Test.php [hunk 2]
```diff
@@ -964,7 +1033,7 @@ public function test_handle_edit_line_item_modal_error_when_payment_not_found():
 		$_REQUEST['payment_id'] = 999999;
 		$_POST['payment_id']    = 999999;
 
-		$this->expectException(\WPDieException::class);
+		$this->expectException(\Error::class);
 
 		$this->page->handle_edit_line_item_modal();
```
[the remark this answers is not in the document]
Comment (superdav42) [reply tests/WP_Ultimo/Admin_Pages/Payment_Edit_Admin_Page_Test.php]: @coderabbitai Addressed in dfc8a34e: `handle_edit_line_item_modal()` now returns `not-found` JSON when the payment is missing, and the test asserts that contract (`inc/admin-pages/class-payment-edit-admin-page.php`, `tests/WP_Ultimo/Admin_Pages/Payment_Edit_Admin_Page_Test.php`). Verified with focused PHPUnit: 303 tests, 564 assertions, 4 skipped.

File: tests/WP_Ultimo/Admin_Pages/Settings_Admin_Page_Test.php [hunk 3]
```diff
@@ -324,13 +490,33 @@ public function test_default_handler_dies_without_permission(): void {
...
+
+		try {
+			$this->page->default_handler();
+		} catch (\RuntimeException $e) {
+			$this->assertSame('redirect_intercepted', $e->getMessage());
+		} finally {
+			remove_all_filters('wp_redirect');
+		}
+
+		$this->assertNotNull($redirect_url, 'wp_redirect should have been called');
```
[the remark this answers is not in the document]
Comment (superdav42) [reply tests/WP_Ultimo/Admin_Pages/Settings_Admin_Page_Test.php:519]: @coderabbitai Addressed in dfc8a34e: redirect interception now removes only its callback and resets current user state in `finally` (`tests/WP_Ultimo/Admin_Pages/Settings_Admin_Page_Test.php`). Verified with focused PHPUnit: 303 tests, 564 assertions, 4 skipped.

File: tests/WP_Ultimo/Admin_Pages/Setup_Wizard_Admin_Page_Test.php [hunk 4]
```diff
@@ -50,6 +51,105 @@ protected function tearDown(): void {
...
+			$this->assertSame('redirect_intercepted', $e->getMessage());
+		} finally {
+			remove_all_filters('wp_redirect');
+		}
+
+		$this->assertNotNull($redirect_url, 'wp_redirect should have been called');
+
+		return $redirect_url;
+	}
+
```
[the remark this answers is not in the document]
Comment (superdav42) [reply tests/WP_Ultimo/Admin_Pages/Setup_Wizard_Admin_Page_Test.php:151]: @coderabbitai Addressed in dfc8a34e: `capture_redirect()` now stores the redirect callback and removes only that callback (`tests/WP_Ultimo/Admin_Pages/Setup_Wizard_Admin_Page_Test.php`). Verified with focused PHPUnit: 303 tests, 564 assertions, 4 skipped.

File: tests/WP_Ultimo/Admin_Pages/Setup_Wizard_Admin_Page_Test.php [hunk 5]
```diff
@@ -357,26 +457,46 @@ public function test_alert_incomplete_installation_returns_early_when_not_loaded
...
+		$_REQUEST['_wpnonce'] = wp_create_nonce('wu_setup_install');
+
+		$response = $this->capture_json_response(
+			function () {
+				$this->page->setup_install();
+			}
+		);
+
+		$this->assertTrue($response['success']);
 	}
```
[the remark this answers is not in the document]
Comment (superdav42) [reply tests/WP_Ultimo/Admin_Pages/Setup_Wizard_Admin_Page_Test.php:490]: @coderabbitai Addressed in dfc8a34e: the privileged setup install test now resets current user state in `finally` (`tests/WP_Ultimo/Admin_Pages/Setup_Wizard_Admin_Page_Test.php`). Verified with focused PHPUnit: 303 tests, 564 assertions, 4 skipped.

File: tests/WP_Ultimo/Admin_Pages/Template_Library_Admin_Page_Test.php [hunk 6]
```diff
@@ -909,8 +941,9 @@ public function test_display_more_info_does_not_throw_when_template_not_found():
 		ob_start();
 		try {
 			$this->page->display_more_info();
-		} catch ( \Exception $e ) {
+		} catch ( \Throwable $e ) {
 			// Some template rendering may throw — that's acceptable.
+			$this->assertInstanceOf( \Throwable::class, $e );
 		}
```
[the remark this answers is not in the document]
Comment (superdav42) [reply tests/WP_Ultimo/Admin_Pages/Template_Library_Admin_Page_Test.php]: @coderabbitai Addressed in dfc8a34e: removed the broad exception-swallowing catches and kept output cleanup in `finally`. The exposed failures were fixed by avoiding `wu_get_template()` path-variable collision and returning early when a template slug is missing (`tests/WP_Ultimo/Admin_Pages/Template_Library_Admin_Page_Test.php`, `inc/functions/template.php`, `inc/admin-pages/class-template-library-admin-page.php`). Verified with focused PHPUnit: 303 tests, 564 assertions, 4 skipped.

File: tests/WP_Ultimo/Gateways/PayPal_OAuth_Handler_Standalone_Test.php [hunk 7]
```diff
@@ -493,68 +563,92 @@ public function test_ajax_disconnect_without_permissions(): void {
...
-	echo json_encode(['success' => false, 'data' => $data]);
+	if (function_exists('\\wp_send_json_error')) {
+		return \wp_send_json_error($data, $status_code);
+	}
+	echo wp_json_encode([
+		'success' => false,
+		'data'    => $data,
+	]);
 	exit;
 }
```
[the remark this answers is not in the document]
Comment (superdav42) [reply tests/WP_Ultimo/Gateways/PayPal_OAuth_Handler_Standalone_Test.php:623]: @coderabbitai Addressed in dfc8a34e: the standalone fallback now uses `wp_json_encode` when available and falls back to `json_encode` otherwise (`tests/WP_Ultimo/Gateways/PayPal_OAuth_Handler_Standalone_Test.php`). Verified with focused PHPUnit and PHPCS on changed production/PayPal files.

**Comment by superdav42** [comment]
CLAIM_RELEASED reason=worker_complete runner=superdav42 ts=2026-06-21T20:32:17Z aidevops_version=3.21.11 opencode_version=1.17.9
